src/constraint_system/boolean.rs

```diff
@@ -57,27 +57,32 @@ where
     pub fn is_zero_gate(&mut self, a: Variable) -> Variable {
         //Get relevant field values
         let a_value = self.variables.get(&a).unwrap();
-        let a_inv_value = a_value.inverse()
-            .unwrap_or(E::Fr::one());
+        let a_inv_value = a_value.inverse().unwrap_or(E::Fr::one());
         //This has value 1 if input value is zero, value 0 otherwise
         let result_value = E::Fr::one() - *a_value * a_inv_value;
 
         let a_inv = self.add_input(a_inv_value);
         let result = self.add_input(result_value);
 
         // Enforce constraints
-        let a_times_result = self.arithmetic_gate(|gate| gate.witness(a, result, None).mul(E::Fr::one()));
+        let a_times_result = self.arithmetic_gate(|gate| {
+            gate.witness(a, result, None).mul(E::Fr::one())
+        });
         self.assert_equal(a_times_result, self.zero_var());
```

Comment:
This `assert_equal` isn't necessary.  The product of `a` and `result` is constrained to be zero by the `mul` gate itself if you use `gate.witness(a, result, self.zero_var())`

Reply:
Thanks for the tip, I see how to use the `w_o` option correctly now.  Unfortunately doing so causes my tests to fail.  The error seems to occur here [https://github.com/arkworks-rs/algebra/blob/7657bcb7280ffb112056d59f2817c484ee4a6a6d/poly/src/polynomial/univariate/dense.rs#L35](https://github.com/arkworks-rs/algebra/blob/7657bcb7280ffb112056d59f2817c484ee4a6a6d/poly/src/polynomial/univariate/dense.rs#L35)

when getting the degree of a polynomial during proof generation.  Any idea why using `w_o : None` and `assert_equal` works but combining them as you suggested wouldn't? The exact change I made is to replace lines 68-70 with   
`let a_times_result = self.arithmetic_gate(|gate| { gate.witness(a, result, Some(zero)).mul(E::Fr::one())});`

where `zero` is `self.zero_var()`  
  
Could the problem be that this creates a variable `a_times_result` which is never used by the circuit?